rho/ansible_utils.py

```diff
@@ -0,0 +1,165 @@
+
+    :returns: a dict that can be used as the host variables in an
+        Ansible inventory.
+    """
+
+    username = auth.get('username')
+    password = auth.get('password')
+    ssh_key_file = auth.get('ssh_key_file')
+    sudo_password = auth.get('sudo_password')
+
+    ansible_vars = {}
+
+    ansible_vars['ansible_user'] = str_to_ascii(username)
+    if password:
+        ansible_vars['ansible_ssh_pass'] = str_to_ascii(password)
+    if ssh_key_file:
+        ansible_vars['ansible_ssh_private_key_file'] = \
+            str_to_ascii(ssh_key_file)
+    if sudo_password:
+        ansible_vars['ansible_become_pass'] = sudo_password
+
+    return ansible_vars
+
+
+def redact_dict(redact_key_list, a_dict):
+    """Redact_values in a dictionary
+
+    :param redact_key_list: keys in dictionary that value should be redacted
+    :param a_dict: A dictionary to redact
+    :returns: the redacted dictionary
+    """
+    for key in redact_key_list:
+        if a_dict is not None and key in a_dict:
+            a_dict[key] = utilities.PASSWORD_MASKING
+    return a_dict
+
+
+def log_yaml_inventory(label, inventory):
+    """Log yaml inventory but mask passwords
+
+    :param inventory: A dictionary of the ansible inventory
+    """
+    alpha = inventory.get('alpha')
+    hosts_dict = alpha.get('hosts')
+    vars_dict = alpha.get('vars')
+    redact_key_list = ['ansible_become_pass', 'ansible_ssh_pass']
+
+    # pylint: disable=unused-variable
+    for host, host_dict in iteritems(hosts_dict):
+        host_dict = redact_dict(redact_key_list, host_dict)
+
+    vars_dict = redact_dict(redact_key_list, vars_dict)
+
+    log.debug('%s:\n%s', label, yaml.dump(inventory))
+    return inventory
+
+
+# pylint: disable=too-many-arguments
+def run_with_vault(cmd_string, vault_pass, env=None, log_path=None,
+                   log_to_stdout=True, ansible_verbosity=0):
+    """Runs ansible command allowing for password to be provided after
+    process triggered.
+
+    Returns after the process completes.
+
+    :param cmd_string: the command to run.
+    :param vault_pass: the password to the user's Ansible Vault.
+    :param env: the environment to run the subprocess in.
+    :param log_path: a path to write the process's log to. Defaults to
+        'XDG_DATA_HOME/rho/ansible_log'.
+    :param log_to_stdout: if True, write Ansible's log to stdout. Defaults to
+        True.
+    :param ansible_verbosity: the number of v's of Ansible verbosity.
+    :returns: the popen.spawn object for the process.
+    """
+
+    # pexpect provides the ability to send the process's output to a
+    # single Python file object. We want to send it to a file and
+    # maybe also stdout. The solution is to have pexpect log to the
+    # file and then use 'tail -f' to copy that to stdout.
+
+    if not log_path:
+        log_path = ANSIBLE_LOG_PATH
+
+    if ansible_verbosity:
+        cmd_string = cmd_string + ' -' + 'v' * ansible_verbosity
```

Comment:
Possibly `cmd_string = '{} -{}'.format(cmd_string, 'v' * ansible_versbosity)`. 

I've been discouraged from this '+' style of string concatenation, but I think its mostly a style issue.

Reply:
In general I do like using `.format()`, but in this case because the format string is so simple, it seems like it adds complexity but doesn't really buy us clarity.